From bp@alien8.de Mon Jul 21 17:45:57 2014
Message-ID: <53BC7130.1010105@gmail.com>
Date:	Tue, 08 Jul 2014 17:31:12 -0500
From: Stuart Hayes <stuart.w.hayes@gmail.com>
To: tglx@linutronix.de, mingo@redhat.com, hpa@zytor.com, x86@kernel.org
CC: linux-kernel@vger.kernel.org, bpetkov@suse.com, matt.fleming@intel.com
Subject: [PATCH] x86: Allow kernel_map_pages_in_pgd() to work when NX is disabled
Patch-mainline: not yet
References: bnc#884369

The function kernel_map_pages_in_pgd() will not map pages if NX is
disabled, which causes a lot of problems booting in EFI mode
(efi_map_region() and other functions depend on this).  This patch
just makes sure that the NX flag doesn't get set in the page tables
if NX is disabled, rather than not mapping the pages at all.

Signed-off-by: Stuart Hayes <stuart.w.hayes@gmail.com>
Link: http://lkml.kernel.org/r/53BC7130.1010105@gmail.com
Acked-by: Borislav Petkov <bp@suse.de>
---
---
 arch/x86/mm/pageattr.c |    6 +-----
 1 file changed, 1 insertion(+), 5 deletions(-)

--- a/arch/x86/mm/pageattr.c
+++ b/arch/x86/mm/pageattr.c
@@ -1862,10 +1862,7 @@ int kernel_map_pages_in_pgd(pgd_t *pgd,
 		.flags = 0,
 	};
 
-	if (!(__supported_pte_mask & _PAGE_NX))
-		goto out;
-
-	if (!(page_flags & _PAGE_NX))
+	if ((!(__supported_pte_mask & _PAGE_NX)) || !(page_flags & _PAGE_NX))
 		cpa.mask_clr = __pgprot(_PAGE_NX);
 
 	cpa.mask_set = __pgprot(_PAGE_PRESENT | page_flags);
@@ -1873,7 +1870,6 @@ int kernel_map_pages_in_pgd(pgd_t *pgd,
 	retval = __change_page_attr_set_clr(&cpa, 0);
 	__flush_tlb_all();
 
-out:
 	return retval;
 }
 
